cmd/kwild/server/migration.go

```diff
@@ -0,0 +1,161 @@
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/kwilteam/kwil-db/common/chain"
+	commonCfg "github.com/kwilteam/kwil-db/common/config"
+	"github.com/kwilteam/kwil-db/core/client"
+	"github.com/kwilteam/kwil-db/core/types"
+	"github.com/kwilteam/kwil-db/internal/statesync"
+)
+
+// The migrationClient type is responsible for:
+//   - Polling the old chain to retrieve the genesis state required for migration.
+//   - Downloading the genesis snapshot from the old chain and saves it under the
+//     root directory of the node.
+//   - Updating the genesis configuration such as app hash, validators and migration settings.
+//   - Updating the kwild configuration with the snapshot file path and migrations listener extension.
+
+const (
+	defaultPollFrequency = time.Second * 30
+)
+
+type migrationClient struct {
+	// listenAddress is the old chain's listen address to retrieve the genesis state
+	listenAddress string
+
+	snapshotFileName string
+
+	clt        *client.Client
+	kwildCfg   *commonCfg.KwildConfig
+	genesisCfg *chain.GenesisConfig
+}
+
+func PrepareForMigration(ctx context.Context, kwildCfg *commonCfg.KwildConfig, genesisCfg *chain.GenesisConfig) (*commonCfg.KwildConfig, *chain.GenesisConfig, error) {
+	if kwildCfg.MigrationConfig.MigrateFrom == "" {
+		return nil, nil, errors.New("migrate_from is mandatory for migration")
+	}
+
+	// old chain client
+	clt, err := client.NewClient(ctx, kwildCfg.MigrationConfig.MigrateFrom, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Get the genesis state from the old chain
+	m := &migrationClient{
+		listenAddress:    kwildCfg.MigrationConfig.MigrateFrom,
+		clt:              clt,
+		kwildCfg:         kwildCfg,
+		genesisCfg:       genesisCfg,
+		snapshotFileName: filepath.Join(kwildCfg.RootDir, "snapshot.sql.gz"),
+	}
+
+	// poll for the genesis state
+	if err = m.pollForGenesisState(ctx); err != nil {
+		return nil, nil, err
+	}
+
+	return m.kwildCfg, m.genesisCfg, nil
+}
+
+// pollForGenesisState polls for the genesis state from the old chain
+// and downloads the genesis state to the snapshot file abnd updates the genesis config
+// and the kwild config with the configuration required for migration
+func (m *migrationClient) pollForGenesisState(ctx context.Context) error {
+	// Poll for the genesis state from the old chain
+	delay := defaultPollFrequency
+	ticker := time.NewTicker(delay)
```

Comment:
nbd but what is the point of copying `defaultPollFrequency` into `delay` just to use it in the next line?